docs/markdown/Writing Plugins/tutorials/01-create-new-goal-with-a-custom-target.md

```diff
@@ -0,0 +1,335 @@
+
+At this point, we are ready to do something useful with the new target of ours. Goals generally run on targets, so they need to be passed as an argument in the command line. For instance, to format the `myproject` directory targets, you would run `./pants fmt myproject`. To get the version of a project in your repository, it makes sense to pass to the `project-version` goal a project directory containing the `version_file` definition.
+
+To make a target passed as an argument accessible in the goal rule, we pass the [`Targets`](https://www.pantsbuild.org/docs/rules-api-goal-rules#how-to-operate-on-targets) as input arguments of the function along with the [`Console`](https://www.pantsbuild.org/docs/rules-api-goal-rules#console-output-to-stdoutstderr) object so that we can print the details of our target in the user terminal:
+
+```python
+@goal_rule
+async def goal_show_project_version(console: Console, targets: Targets) -> ProjectVersionGoal:
+    # since we don't know what targets will be passed (e.g. `::`), we want to keep only `version_file` targets
+    targets = [tgt for tgt in targets if tgt.alias == ProjectVersionTarget.alias]
+    for target in targets:
+        console.print_stdout(target.address.metadata())
+    return ProjectVersionGoal(exit_code=0)
+
+```
+
+Having the following directory structure:
+
+```text
+myapp
+├── BUILD
+└── VERSION
+```
+
+we are ready to inspect our new target:
+
+```text
+$ ./pants project-version myapp
+{'address': 'myapp:main-project-version'}
+{'includes': ['myapp/VERSION']}
+```
+
+## Writing a rule
+
+You can think of the `@goal_rule` as of the `main` function in your Python program where you would call various functions that your program needs to complete. For auxiliary code, it makes sense to place it into standalone functions which is what `@rule`s are for.
+
+Let's create a rule that will return a data structure that we'll use to represent our project version. [Data classes](https://www.pantsbuild.org/docs/rules-api-concepts#dataclasses) work really well with Pants engine, so let's create one: 
+
+```python
+@dataclass(frozen=True)
+class ProjectVersionFileView:
+    path: str
+    version: str
+```
+
+This is what our `@rule` function would return (for now without actually reading the `VERSION` file):
+
+```python
+@rule
+async def get_project_version_file_view(
+    target: ProjectVersionTarget,
+) -> ProjectVersionFileView:
+    return ProjectVersionFileView(
+        path="path", version="1.2.3"
+    )
+```
+
+Now, we have our `@goal_rule`, but we cannot call the `get_project_version_file_view` function; it's Pants that will determine that a rule is used and will make a function call. Well, what should you do to tell Pants you need that rule executed? You should make a function call that:
+* passes an object of the type that matches the type of the rule's input arguments
+* requests an object of the type that a rule returns (you can see that in a type hint)
```

Comment:
I think it is worth phrasing this so it becomes clear that you provide some input type, and ask for some other output type, and then Pants looks at all the rules in the graph to find a way going from input to output using all of the available rules as necessary.

That is, it's not always one request one rule as in this (simple) example.

Reply:
That's brilliant, I've added this!